[AMDGPU][SIInsertWaitcnts] drop OldWaitcntInstr only when it is processed

ssahasra opened the pull request:
When iterating over instructions, the pass keeps track of pending soft waitcnt
instructions. These are usually processed when a non-waitcnt instruction is
visited, but some visit may return early. Dropping the recorded pointer at this
early exit produces spurious waitcnt. This does not affect correctness, but the
waitcnt is unnecessary.

To fix that, the recorded pointer should be dropped iff pending soft waitcnt
instructions have been processed.

ssahasra committed:
pre-commit test for trailing waitcnt

ssahasra committed:
[AMDGPU][SIInsertWaitcnts] drop OldWaitcntInstr only when it is processed

When iterating over instructions, the pass keeps track of pending soft waitcnt
instructions. These are usually processed when a non-waitcnt instruction is
visited, but some visit may return early. Dropping the recorded pointer at this
early exit produces spurious waitcnt. This does not affect correctness, but the
waitcnt is unnecessary.

To fix that, the recorded pointer should be dropped iff pending soft waitcnt
instructions have been processed.

jmmartinez commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -2490,7 +2493,6 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
     // Generate an s_waitcnt instruction to be placed before Inst, if needed.
     Modified |= generateWaitcntInstBefore(Inst, ScoreBrackets, OldWaitcntInstr,
                                           FlushVmCnt);
-    OldWaitcntInstr = nullptr;
```
If I follow everything, the main case where assigning `OldWaitcntInstr` to `nullptr` here was problematic was for meta-instructions ? Or is there another case ?

If that's the case, we could skip meta-instructions directly in the this loop.

ssahasra replied:
I don't see how we can do that. When we enter `generateWaitcntInstBefore()`, `OldWaitcntInstr` is simply the first such instruction that was found, waiting to be included in a computation. The call actually processes a range starting from `OldWaitcntInstr` to `Inst`. See the `for` loop inside `applyPreexistingWaitcnt()`. It's not enough to just check `Inst` or `OldWaitcntInstr` at this point in the code.

jmmartinez replied:
Let me explain what I meant.
- The range between `OldWaitcntInstr` and `Inst` is a range of waits. When we call `generateWaitcntInstBefore`, `Inst` is the first instruction not to be a wait. In `generateWaitcntInstBefore` we call `applyPreexistingWaitcnt` to try to simplify and make the waits "non-soft".
- Before this patch, a meta-instruction used to "break" the sequence of waits between `OldWaitcntInstr` and `Inst` by resetting `OldWaitcntInstr` to `nullptr`. If I understood, this is the main issue this patch is addressing. `generateWaitcntInstBefore` on a meta-instruction just returns `false`.
- With this patch, `generateWaitcntInstBefore` still doesn't do anything on meta-instructions, and we reset `OldWaitcntInstr` to `nullptr` only when we call `applyPreexistingWaitcnt`.

My questions are:
1. Did I understand the problem this patch is trying to solve correctly ?
2. Is there some other case besides meta-instructions where we used to reset `OldWaitcntInstr` when it did not make sense ?

With this in mind, I think that we can try to pull the condition below out of `generateWaitcntInstBefore` and into the loop.

From:
```cpp
  // in generateWaitcntInstBefore
  if (MI.isMetaInstruction())
    return false;
```

To;
```cpp
// in the loop before "if (isWaitInstr(Inst))"
if (MI.isMetaInstruction()) {
  ++Iter;
  continue;
}
```

If we do this, we could keep the unconditional `OldWaitcntInstr = nullptr` after `generateWaitcntInstBefore`. And avoid having this assignment deep in `generateWaitcnt`.

ssahasra committed:
[AMDGPU][SIInsertWaitCnts] skip meta instructions early

When iterating over a block, meta instructions have no effect on wait counts,
but their presence drops the reference to earlier waitcnt instructions before
they are processed. This results in spurious wait counts, which do not affect
correctness, but are also not required in the resulting program. Skipping meta
instructions as soon as they are seen cleans this up.

jmmartinez reviewed (approved):
Thanks! Everything else looks good to me.

ssahasra committed:
replace check for meta instruction with an assert

ssahasra committed:
cleanup CHECK lines in the new test
